interop-scoring/main.js

```diff
@@ -0,0 +1,506 @@
+          // in the subtests object to determine interop score for the test.
+          if (!(interopScores.get(testname).has('overall'))) {
+            interopScores.get(testname).set('overall', 0);
+          }
+          if (results['status'] == 'PASS') {
+            subtestPasses = 1;
+            const previousVal = interopScores.get(testname).get('overall');
+            interopScores.get(testname).set('overall', previousVal + 1);
+          }
+        }
+        // A single test is scored 0-1000 based on how many of its subtests
+        // pass, rounding down so that 1000 always means fully passing.
+        score += Math.floor(1000 * subtestPasses / subtestTotal);
+      });
+      // We always normalize against the number of tests we are looking for,
+      // rather than the total number of tests we found. The trade-off is all
+      // about new tests being added to the set.
+      //
+      // If a large chunk of tests are introduced at date X, and they fail in
+      // some browser, then runs after date X look worse if you're only
+      // counting total tests found - even though the tests would have failed
+      // before date X as well.
+      //
+      // Conversely, if a large chunk of tests are introduced at date X, and
+      // they pass in some browser, then runs after date X would get an
+      // artificial boost in pass-rate due to this - even if the tests would
+      // have passed before date X as well.
+      //
+      // We consider the former case worse than the latter, so optimize for it
+      // by always comparing against the full test list. This does mean that
+      // when tests are added to the set, previously generated data is no
+      // longer valid and this script should be re-run for all dates.
+      scores.push(Math.floor(score / allTestsSet.size));
+    }
+  } catch (e) {
+    e.message += `\n\tRuns: ${runs.map(r => r.id)}`;
+    throw e;
+  }
+  // Calculate the interop scores that have been saved and add
+  // The interop score to the end of the browsers' scores array.
+  scores.push(aggregateInteropTestScores(interopScores, runs.length));
+  return scores;
+}
+
+async function scoreCategory(category, experimental, products, alignedRuns,
+    testsSet) {
+  // Score the test runs.
+  const before = Date.now();
+  const dateToScores = new Map();
+  for (const [date, runs] of alignedRuns.entries()) {
+    const versions = runs.map(run => run.browser_version);
+    const scores = scoreRuns(runs, testsSet);
+    dateToScores.set(date, {versions, scores});
+  }
+  const after = Date.now();
+  console.log(`Done scoring (took ${after - before} ms)`);
+
+  // Return dateToScores, so that our caller can calculate the summary across
+  // multiple categories.
+  return dateToScores;
+}
+
+async function main() {
+  const year = (flags.isSet('year')) ? flags.get('year') : '2023';
+  if (!year in interopData) {
+    throw new Error(`Categories not defined for year ${year}`);
+  }
```

Comment:
The `--year` flag is now available, and will default to 2023 if not set. Additionally, the script will fail if invoked on a year that does not have categories explicitly defined in category-data.json.